spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -102,12 +115,39 @@ public void onMessage(org.springframework.amqp.core.Message amqpMessage, Channel
 		if (logger.isDebugEnabled()) {
 			logger.debug("Processing [" + message + "]");
 		}
-		Object result = invokeHandler(amqpMessage, channel, message);
-		if (result != null) {
-			handleResult(result, amqpMessage, channel, message);
+		try {
+			Object result = invokeHandler(amqpMessage, channel, message);
+			if (result != null) {
+				handleResult(result, amqpMessage, channel, message);
+			}
+			else {
+				logger.trace("No result object given - no result to handle");
+			}
 		}
-		else {
-			logger.trace("No result object given - no result to handle");
+		catch (ListenerExecutionFailedException e) {
+			if (this.errorHandler != null) {
+				try {
+					Object result = this.errorHandler.handleError(amqpMessage, message, e);
+					if (result != null) {
+						handleResult(result, amqpMessage, channel, message);
+					}
+					else {
+						logger.trace("Error handler returned no result");
+					}
+				}
+				catch (Exception ex) {
+					if (!this.propagateExceptions) {
+						throw ex;
+					}
+					handleResult(new RemoteInvocationResult(ex), amqpMessage, channel, message);
```

Comment:
Would you mind explaining the reason of the `RemoteInvocationResult` layer?
To get a gain from the `recreate()` ability to rethrow?
How about just to equip `SimpleMessageConverter` with a flag `rethrowResult`, since it looks like with this feature we always can deserialize `Exception` as is?